core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,539 @@
+	maxBackoffDelay time.Duration
+	// The block censorship timeout. A block censorship suspicion is declared if more than f header receivers are
+	// ahead of the block receiver for a period larger than this timeout.
+	blockCensorshipTimeout time.Duration
+
+	endpoints          []comm.EndpointCriteria // a set of endpoints
+	blockReceiverIndex int                     // index of the current block receiver endpoint
+
+	blockReceiver   *broadcastClient
+	nextBlockNumber uint64
+	lastBlockTime   time.Time
+
+	headerReceivers map[string]*bftHeaderReceiver
+}
+
+func NewBFTDeliveryClient(
+	chainID string,
+	connFactory comm.ConnectionFactory,
+	endpoints []comm.EndpointCriteria,
+	clFactory clientFactory,
+	ledgerInfoProvider blocksprovider.LedgerInfo,
+	msgVerifier MessageCryptoVerifier,
+) *bftDeliveryClient {
+	c := &bftDeliveryClient{
+		stopChan:                            make(chan struct{}, 1),
+		chainID:                             chainID,
+		connectionFactory:                   connFactory,
+		createClient:                        clFactory,
+		ledgerInfoProvider:                  ledgerInfoProvider,
+		msgCryptoVerifier:                   msgVerifier,
+		reconnectTotalTimeThreshold:         getReConnectTotalTimeThreshold(),
+		reconnectBlockRcvTotalTimeThreshold: util.GetDurationOrDefault("peer.deliveryclient.bft.blockRcvTotalBackoffDelay", bftBlockRcvTotalBackoffDelay),
+		minBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.minBackoffDelay", bftMinBackoffDelay),
+		maxBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.maxBackoffDelay", bftMaxBackoffDelay),
+		blockCensorshipTimeout:              util.GetDurationOrDefault("peer.deliveryclient.bft.blockCensorshipTimeout", bftBlockCensorshipTimeout),
+		endpoints:                           comm.Shuffle(endpoints),
+		blockReceiverIndex:                  -1,
+		headerReceivers:                     make(map[string]*bftHeaderReceiver),
+	}
+
+	bftLogger.Infof("[%s] Created BFT Delivery Client", chainID)
+	return c
+}
+
+func (c *bftDeliveryClient) Recv() (response *orderer.DeliverResponse, err error) {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	c.startOnce.Do(func() {
+		var num uint64
+		num, err = c.ledgerInfoProvider.LedgerHeight()
+		if err != nil {
+			return
+		}
+		c.nextBlockNumber = num
+		c.lastBlockTime = time.Now()
+		bftLogger.Debugf("[%s] Starting monitor routine; Initial ledger height: %d", c.chainID, num)
+		go c.monitor()
+	})
+	// can only happen once, after first invocation
+	if err != nil {
+		bftLogger.Debugf("[%s] Exit: error=%v", c.chainID, err)
+		return nil, errors.New("cannot access ledger height")
+	}
+
+	var numEP int
+	var numRetries int
+	var stopRetries = time.Now().Add(c.reconnectTotalTimeThreshold)
+
+	for !c.shouldStop() {
+		if numEP, err = c.assignReceivers(); err != nil {
+			bftLogger.Debugf("[%s] Exit: error=%s", c.chainID, err)
+			return nil, err
+		}
+
+		c.launchHeaderReceivers()
```

Comment:
this function is within a for loop, that might return right below, and it launches **f** goroutines. What if we do it 2 or more times in a row? won't we have parallel Recv() called ? The base client was never designed to handle parallel receives. 

I think this whole process should be better managed and we should never launch goroutines in the green path, but only at initialization or whenever we have a suspicion of censorship.

Reply:
discussed f2f